### ISSUE TITLE
Remove tensorframes dependency

### ISSUE THREAD

**Issue by davidagold** [mapbox/trace-classifier#17]
Tensorframes is deprecated and we need to remove the dependency here in order to be able to integrate into trace-maker.

cc: @kierangupta

**Comment by davidagold** [comment]
@lily-chai Do you recall the details concerning your saving the serialized model?

We are working on removing the dependency on tensorframes and in doing so need to apply the Keras model via (Pandas) UDF as in [this tutorial](https://docs.databricks.com/applications/deep-learning/inference/resnet-model-inference-keras.html).

To do so, we need to be able to import the serialized model as a `tf.keras.Model` object. However, when attempting to load the model through [`tf.keras.models.load_model`](https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/keras/models/load_model), the result is an `AutoTrackable` object with no `__call__` or `predict` methods and we get a warning:
```
INFO:tensorflow:Saver not created because there are no variables in the graph to restore
```

I am assuming this is because there is no `variables` directory available (as noted in the [serialization docs](https://www.tensorflow.org/guide/keras/save_and_serialize)) which I in turn expect is because the `.pb` that is available was not serialized via `tf.keras.models.save_model` but rather through https://github.com/mapbox/trace-classifier/blob/67b46cf8f1f36487525ae68d85e217116052c31c/trace_classifier/save.py#L104-L159. 

So, my two specific questions are:
1. Is there a way to convert the frozen graph `.pb` file to a `tf.keras.models.load`-able file?
1. If not, is there a way to serve the former using a Pandas UDF?
1. If not, what is the recourse? A (somewhat radical) fallback seems to be to re-train the model and save the full model information using the `tf.keras.models.save` API rather than the frozen graph method. 

Also, the above has been with tensorflow upgraded to 1.15.0, (from 1.10.0 in the current trace-classifier version).

@ishivvers do you have much familiarity with serialized keras models/tensorflow graphs?

cc @sabaelhilo @tcql

**Comment by ishivvers** [comment]
@davidagold : some - just enough to be concerned by this 🙃   What I've learned from doing this sort of work in the past is that these deep learning libraries are under such constant (and complex) development that it's more-or-less a full time job to keep your models up to date with the underlying developments.

Some thoughts / comments:

- The inter-relationship between `keras` and `tensorflow` has changed dramatically over the last 2 years: with TF2.0 the `keras` API was officially folded in as part of `tensorflow`, but for previous versions the true `keras` was a standalone third-party wrapper around the `tensorflow` API (and other APIs), while the `tensorflow` team kindof also had an integrated version of the `keras` API called `tensorflow.keras` (e.g., see [here](https://github.com/keras-team/keras/releases/tag/2.4.0)).  Basically it was just a mess for a while, and it probably will not be straightforward to take a pre-2.0 `tensorflow` computation graph and convert it into a `keras` model; there will be a lot of additional `keras`-specific metadata missing, I'm betting, and it probably matters _which flavor of keras was used / is now used_.

- My initial gut reaction is that your somewhat radical proposal of re-training the model is easier in the long run - I would advocate for updating the model to be implemented in TF2.0+Keras and then re-training.  Maintaining a TF1.0+Keras model is only going to get more difficult as things continue to change and the various workarounds required get lost to the sands of time.

**Comment by davidagold** [comment]
Thanks for your input @ishivvers! That's super helpful. 

>My initial gut reaction is that your somewhat radical proposal of re-training the model is easier in the long run - I would advocate for updating the model to be implemented in TF2.0+Keras and then re-training. Maintaining a TF1.0+Keras model is only going to get more difficult as things continue to change and the various workarounds required get lost to the sands of time.

That makes sense. In that case I believe the steps required to pursue this course of action are:

- [ ] Locate trace-classifier tickets containing model specification, training information (verify as much as possible that the repository code for training reflects what was done to obtain previous model)
- [ ] Upgrade to tensorflow 2.x in trace-classifier (resolve packaging incompatibilities, etc.) 
- [ ] Make any needed changes to [model architecture implementation](https://github.com/mapbox/trace-classifier/blob/67b46cf8f1f36487525ae68d85e217116052c31c/trace_classifier/architecture.py#L28-L50)
- [ ] Fit model against training data from s3://mapbox-modality/datasets/
- [ ] QA fitted model (verify that statistical behavior of predictions is comparable to what we observed in previous QA)
- [ ] Upgrade `trace_classifier.infer.infer` to use [this DataBricks idiom](https://docs.databricks.com/applications/deep-learning/inference/resnet-model-inference-keras.html) 

We should also put the traces in `s3://mapbox-modality/datasets/` in a protected table in the DWH, these are quite valuable and it would be really unfortunate if something happened to them.

**Comment by davidagold** [comment]
@ishivvers One more thing -- while we're doing this work to upgrade, is it worth considering here also whether we want to continue using TensorFlow? I ask because elsewhere at Mapbox -- e.g. Imagery -- we use PyTorch. Does it make sense to do some work to consolidate around a single framework? Do you have suggestions for which framework you think makes the most sense?

**Comment by ishivvers** [comment]
@davidagold : yeah, your checklist looks reasonable to me, if we decide to go that way.

As for PyTorch vs. TensorFlow - I've used TensorFlow more in the past, but I think both libraries are similarly well-supported and powerful.  My understanding is that the complexity of PyTorch's API is somewhere between the (novice-friendly, limited customization possible) Keras API and the (unfriendly, fully customizable) TensorFlow API.

If other parts of the company have settled on PyTorch as the chosen tool set, I'd say we should follow suit, and hopefully we can share some of their tooling / learn from their experience with deployment issues 👍

**Comment by davidagold** [comment]
>If other parts of the company have settled on PyTorch as the chosen tool set, I'd say we should follow suit, and hopefully we can share some of their tooling / learn from their experience with deployment issues

Sounds good, though now that I think about this maybe it is biting off a bit more than I can chew right here and now. Perhaps first things first should just be getting this working without tensorframes and leveraging as much of the infrastructure as we already have. PyTorch can be a next step once @kierangupta is unblocked classifier-related items.

This [late stage experimentation](https://github.com/mapbox/modality-classification/issues/52#issuecomment-421534332) in the model development describes an identical architecture to that implemented in `architecture.py`, which is good -- gives us reason to expect that the extant training code will produce a model that can replicate previous performance.

**Comment by davidagold** [comment]
Currently fixing up aspects of the training process I broke with the previous refactor. Work is being conducted in a local notebook. Training appears to be successful, though there are a couple of discrepancies between the current training data stats and details from reference tickets/notebooks. I have an idea of what is causing this, will investigate. Next step is to set up the model saving/loading and incorporate the DataBricks idiom. I'm OOO tomorrow, will pick this back up on Monday.

**Comment by sabaelhilo** [comment]
@davidagold does this mean the tensorflow refactor is done? is there a pr we should start looking at?

**Comment by davidagold** [comment]
@sabaelhilo Apologies, there is no PR ready yet, and I am missing my guidance of having one ready this week. The PR will not be ready until the final checkmark 

>Upgrade trace_classifier.infer.infer to use this DataBricks idiom

above is complete. The fourth item is mostly complete -- the incomplete part is that I believe there are some "Driving" observations that are not included in the comprehensive training dataset I made the last time around and I would like to try to incorporate them if I am correct. That is the discrepancy I noted above.

**Comment by davidagold** [comment]
Also:
- [ ] train new sample model
